File: Ex03/Java/demo/src/main/java/com/example/ClienteDAO.java
package com.example;

import org.hibernate.Session;
import org.hibernate.Transaction;
import java.util.List;

public class ClienteDAO {

    // === CLIENTE ===
    public void inserir(Cliente cliente) {
        Transaction tx = null;
        try (Session session = HibernateUtil.getSessionFactory().openSession()) {
            tx = session.beginTransaction();
            session.persist(cliente);
            tx.commit();
            System.out.println("Cliente inserido com sucesso!");
        } catch (Exception e) {
            if (tx != null) tx.rollback();
            System.out.println("Erro ao inserir cliente: " + e.getMessage());
        }
    }

    public Cliente buscarPorId(int id) {
        try (Session session = HibernateUtil.getSessionFactory().openSession()) {
            return session.get(Cliente.class, id);
        }
    }

    public List<Cliente> listar() {
        try (Session session = HibernateUtil.getSessionFactory().openSession()) {
            return session.createQuery("from Cliente", Cliente.class).list();
        }
    }

    public void atualizar(Cliente cliente) {
        Transaction tx = null;
        try (Session session = HibernateUtil.getSessionFactory().openSession()) {
            tx = session.beginTransaction();
            session.merge(cliente);
            tx.commit();
            System.out.println("Cliente atualizado com sucesso!");
        } catch (Exception e) {
            if (tx != null) tx.rollback();
            System.out.println("Erro ao atualizar cliente: " + e.getMessage());
        }
    }

    public void remover(int id) {
        Transaction tx = null;
        try (Session session = HibernateUtil.getSessionFactory().openSession()) {
            tx = session.beginTransaction();
            Cliente c = session.get(Cliente.class, id);
            if (c != null) session.remove(c);
            tx.commit();
            System.out.println("Cliente removido com sucesso!");
        } catch (Exception e) {
            if (tx != null) tx.rollback();
            System.out.println("Erro ao remover cliente: " + e.getMessage());
        }
    }

    // === PEDIDO ===
    public void inserirPedido(int clienteId, Pedido pedido) {
        Transaction tx = null;
        try (Session session = HibernateUtil.getSessionFactory().openSession()) {
            tx = session.beginTransaction();
            Cliente c = session.get(Cliente.class, clienteId);
            if (c == null) {
                System.out.println("Cliente não encontrado.");
                return;
            }
            c.addPedido(pedido);
            session.merge(c);
            tx.commit();
            System.out.println("Pedido adicionado com sucesso!");
        } catch (Exception e) {
            if (tx != null) tx.rollback();
            System.out.println("Erro ao adicionar pedido: " + e.getMessage());
        }
    }

    public List<Pedido> listarPedidos() {
        try (Session session = HibernateUtil.getSessionFactory().openSession()) {
            return session.createQuery("from Pedido", Pedido.class).list();
        }
    }

    public Pedido buscarPedidoPorId(int id) {
        try (Session session = HibernateUtil.getSessionFactory().openSession()) {
            return session.get(Pedido.class, id);
        }
    }

    public void atualizarPedido(Pedido pedido) {
        Transaction tx = null;
        try (Session session = HibernateUtil.getSessionFactory().openSession()) {
            tx = session.beginTransaction();
            session.merge(pedido);
            tx.commit();
            System.out.println("Pedido atualizado com sucesso!");
        } catch (Exception e) {
            if (tx != null) tx.rollback();
            System.out.println("Erro ao atualizar pedido: " + e.getMessage());
        }
    }

    public void removerPedido(int pedidoId) {
        Transaction tx = null;
        try (Session session = HibernateUtil.getSessionFactory().openSession()) {
            tx = session.beginTransaction();
            Pedido p = session.get(Pedido.class, pedidoId);
            if (p != null) session.remove(p);
            tx.commit();
            System.out.println("Pedido removido com sucesso!");
        } catch (Exception e) {
            if (tx != null) tx.rollback();
            System.out.println("Erro ao remover pedido: " + e.getMessage());
        }
    }
}

File: Ex03/Python/ClienteDAO.py
from sqlalchemy import create_engine, select
from sqlalchemy.orm import sessionmaker, joinedload
from Cliente import Cliente, Base
from Pedido import Pedido

class ClienteDAO:
    def __init__(self, host, user, password, database):
        db_url = f"mysql+mysqlconnector://{user}:{password}@{host}/{database}"
        self.engine = create_engine(db_url)
        Base.metadata.create_all(self.engine)
        self.Session = sessionmaker(bind=self.engine)

    # === Clientes ===
    def inserir_cliente(self, cliente: Cliente):
        s = self.Session()
        try:
            s.add(cliente); s.commit()
            print(f"Cliente '{cliente.nome}' inserido com ID: {cliente.id}")
            return cliente.id
        except Exception as e:
            s.rollback(); print("Falha ao inserir cliente:", e); return None
        finally:
            s.close()

    def listar_clientes(self):
        s = self.Session()
        try:
            return s.execute(select(Cliente).options(joinedload(Cliente.pedidos))).scalars().all()
        finally:
            s.close()

    def recuperar_por_id_cliente(self, id: int):
        s = self.Session()
        try:
            return s.get(Cliente, id)
        finally:
            s.close()

    def atualizar_cliente(self, cliente: Cliente):
        s = self.Session()
        try:
            s.merge(cliente); s.commit(); print("Cliente atualizado."); return True
        except Exception as e:
            s.rollback(); print("Falha ao atualizar cliente:", e); return False
        finally:
            s.close()

    def remover_cliente(self, id: int):
        s = self.Session()
        try:
            c = s.get(Cliente, id)
            if not c: print("Cliente não encontrado."); return False
            s.delete(c); s.commit(); print("Cliente removido."); return True
        except Exception as e:
            s.rollback(); print("Falha ao remover cliente:", e); return False
        finally:
            s.close()

    # === Pedidos ===
    def inserir_pedido(self, cliente_id: int, pedido: Pedido):
        s = self.Session()
        try:
            c = s.get(Cliente, cliente_id)
            if not c: print("Cliente não encontrado."); return None
            c.pedidos.append(pedido); s.add(c); s.commit()
            print(f"Pedido inserido com ID: {pedido.id}")
            return pedido.id
        except Exception as e:
            s.rollback(); print("Falha ao inserir pedido:", e); return None
        finally:
            s.close()

    def listar_pedidos(self):
        s = self.Session()
        try:
            return s.execute(select(Pedido)).scalars().all()
        finally:
            s.close()

    def recuperar_por_id_pedido(self, id: int):
        s = self.Session()
        try:
            return s.get(Pedido, id)
        finally:
            s.close()

    def atualizar_pedido(self, pedido: Pedido):
        s = self.Session()
        try:
            s.merge(pedido); s.commit(); print("Pedido atualizado."); return True
        except Exception as e:
            s.rollback(); print("Falha ao atualizar pedido:", e); return False
        finally:
            s.close()

    def remover_pedido(self, id: int):
        s = self.Session()
        try:
            p = s.get(Pedido, id)
            if not p: print("Pedido não encontrado."); return False
            s.delete(p); s.commit(); print("Pedido removido."); return True
        except Exception as e:
            s.rollback(); print("Falha ao remover pedido:", e); return False
        finally:
            s.close()
